newIDE/app/src/ObjectEditor/Editors/SpineEditor.js

```diff
@@ -0,0 +1,526 @@
+        // to be done.
+        setTimeout(() => {
+          if (scrollView.current) {
+            scrollView.current.scrollToBottom();
+          }
+        }, 100); // A few ms is enough for a new render to be done.
+      } else {
+        showAlert({
+          title: t`No new animation`,
+          message: t`Every animation from the Spine file is already in the list.`,
+        });
+      }
+    },
+    [
+      forceUpdate,
+      skeleton,
+      spineConfiguration,
+      onObjectUpdated,
+      onSizeUpdated,
+      showAlert,
+    ]
+  );
+
+  const addAnimation = React.useCallback(
+    () => {
+      setNameErrors({});
+
+      const emptyAnimation = new gd.SpineAnimation();
+      spineConfiguration.addAnimation(emptyAnimation);
+      emptyAnimation.delete();
+      forceUpdate();
+      onSizeUpdated();
+      if (onObjectUpdated) onObjectUpdated();
+
+      // Scroll to the bottom of the list.
+      // Ideally, we'd wait for the list to be updated to scroll, but
+      // to simplify the code, we just wait a few ms for a new render
+      // to be done.
+      setTimeout(() => {
+        if (scrollView.current) {
+          scrollView.current.scrollToBottom();
+        }
+      }, 100); // A few ms is enough for a new render to be done.
+    },
+    [forceUpdate, onObjectUpdated, onSizeUpdated, spineConfiguration]
+  );
+
+  const removeAnimation = React.useCallback(
+    animationIndex => {
+      setNameErrors({});
+
+      spineConfiguration.removeAnimation(animationIndex);
+      forceUpdate();
+      onSizeUpdated();
+      if (onObjectUpdated) onObjectUpdated();
+    },
+    [forceUpdate, onObjectUpdated, onSizeUpdated, spineConfiguration]
+  );
+
+  const moveAnimation = React.useCallback(
+    (targetIndex: number) => {
+      const draggedIndex = draggedAnimationIndex.current;
+      if (draggedIndex === null) return;
+
+      setNameErrors({});
+
+      spineConfiguration.moveAnimation(
+        draggedIndex,
+        targetIndex > draggedIndex ? targetIndex - 1 : targetIndex
+      );
+      forceUpdate();
+    },
+    [spineConfiguration, forceUpdate]
+  );
+
+  const changeAnimationName = React.useCallback(
+    (animationIndex, newName) => {
+      const currentName = spineConfiguration
+        .getAnimation(animationIndex)
+        .getName();
+      if (currentName === newName) return;
+      const animation = spineConfiguration.getAnimation(animationIndex);
+
+      setNameErrors({});
+
+      if (newName !== '' && spineConfiguration.hasAnimationNamed(newName)) {
+        // The indexes can be used as a key because errors are cleared when
+        // animations are moved.
+        setNameErrors({
+          ...nameErrors,
+          [animationIndex]: (
+            <Trans>The animation name {newName} is already taken</Trans>
+          ),
+        });
+        return;
+      }
+
+      animation.setName(newName);
+      // TODO EBO Refactor event-based object events when an animation is renamed.
```

Comment:
Do todo

Reply:
it is GDevelop author TODO

Reply:
In totally new file?

Reply:
SpineEditor made by Model3DEditor template. Similar functionality, similar problems